vault/resource_config_ui_custom_message.go

```diff
@@ -0,0 +1,277 @@
+			},
+			consts.FieldType: {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "banner",
+				ValidateDiagFunc: func(value interface{}, _ cty.Path) diag.Diagnostics {
+					stringValue := value.(string)
+					switch {
+					case stringValue != "banner" && stringValue != "modal":
+						return diag.Diagnostics{diag.Diagnostic{
+							Severity: diag.Error,
+							Summary:  "invalid value for \"type\" argument",
+							Detail:   "The \"type\" argument can only be set to \"banner\" or \"modal\".",
+						}}
+					}
+
+					return nil
+				},
+				Description: "The display type of custom message. Allowed values are banner and modal",
+			},
+			consts.FieldStartTime: {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "The starting time of the active period of the custom message",
+			},
+			consts.FieldEndTime: {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "The ending time of the active period of the custom message. Can be omitted for non-expiring message",
+			},
+			consts.FieldLink: {
+				Type:     schema.TypeSet,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"title": {
+							Type:        schema.TypeString,
+							Required:    true,
+							Description: "The title of the hyperlink",
+						},
+						"href": {
+							Type:        schema.TypeString,
+							Required:    true,
+							Description: "The URL of the hyperlink",
+						},
+					},
+				},
+				Description: "A block containing a hyperlink associated with the custom message",
+			},
+			consts.FieldOptions: {
+				Type:        schema.TypeMap,
+				Optional:    true,
+				Description: "A map containing additional options for the custom message",
+			},
+		},
+	}
+}
+
+func configUICustomMessageCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	if !provider.IsEnterpriseSupported(meta) {
+		return diag.Errorf("config_ui_custom_message is not supported by this version of vault")
+	}
+
+	client, e := provider.GetClient(d, meta)
+	if e != nil {
+		return diag.FromErr(e)
+	}
+
+	secret, e := client.Sys().CreateUICustomMessageWithContext(ctx, buildUICustomMessageRequest(d))
+	if e != nil {
+		return diag.FromErr(e)
+	}
+
+	if secret == nil || secret.Data == nil {
+		return diag.Errorf(`response from Vault server is empty`)
+	}
+
+	id, ok := secret.Data[consts.FieldID]
+	if !ok {
+		return diag.Errorf("error creating custom message: %s", secret.Data["error"])
+	}
+
+	d.SetId(id.(string))
+
+	return configUICustomMessageRead(ctx, d, meta)
+}
+
+func configUICustomMessageRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, e := provider.GetClient(d, meta)
+	if e != nil {
+		return diag.FromErr(e)
+	}
+
+	id := d.Id()
+
+	log.Printf("[DEBUG] Reading custom message %q", id)
+	secret, e := client.Sys().ReadUICustomMessage(id)
+	if e != nil {
```

Comment:
Does this endpoint return a 404 error if the resource isn't found or is the next if block how we know the resource does not exist in Vault? If Vault does return a 404, we should remove it from TF state (`d.setId("")`) here as well.

Reply:
Here is an example of doing this check: 

https://github.com/hashicorp/terraform-provider-vault/blob/9f9782413129b86df4c301ce356dd6934e5e11a6/vault/resource_approle_auth_backend_role.go#L299-L307

Reply:
Vault does indeed return an HTTP 404 error if no message with the ID provided to the **ReadUICustomMessage** function exists.